### PR TITLE
use discord-markdown and adapt everything to it

### PR DESCRIPTION
**Pull request by Sorunome**
This switches away from `marked` to a custom version of `discord-markdown` ( https://github.com/Sorunome/discord-markdown ), fixing a lot of markdown / formatting issues:

This includes:
* #305 (fixing #272 and #289)
* #304 (implementing #196)
* fixing #55 
* fixing #222

### CODE REVIEW

File: src/messageprocessor.ts [hunk 1]
```diff
@@ -56,6 +33,24 @@ export class MessageProcessorMatrixResult {
     public msgtype: string;
 }
 
+interface IUserNode {
```
Comment (Half-Shot) [inline src/messageprocessor.ts]: I feel like we could just call this IDiscordMsgNode and have just IEnojiNode inherit it, full OOP style.

File: src/messageprocessor.ts [hunk 2]
```diff
@@ -131,111 +128,136 @@ export class MessageProcessor {
...
+            content = content.replace(results[0],
+                replace);
+            results = MXC_INSERT_REGEX.exec(content);
         }
         return content;
     }
-}
 
-_setupMarked();
+    private getDiscordCallback(msg: Discord.Message) {
```
[the remark this answers is not in the document]
Comment (Sorunome) [reply src/messageprocessor.ts]: yes, but then soru would have had to copypaste it twice, once for embeds and once for the normal part

File: test/test_messageprocessor.ts [hunk 3]
```diff
@@ -42,7 +43,7 @@ describe("MessageProcessor", () => {
             msg.content = "Hello *World*!";
             const result = await processor.FormatDiscordMessage(msg);
             Chai.assert.equal(result.body, "Hello *World*!");
-            Chai.assert.equal(result.formattedBody, "<p>Hello <em>World</em>!</p>");
+            Chai.assert.equal(result.formattedBody, "Hello <em>World</em>!");
```
[the remark this answers is not in the document]
Comment (Sorunome) [reply test/test_messageprocessor.ts]: it appears discord doesn't do this "properly", either:

```
test
\`\`\`
multiline
code
\`\`\`
ing
```

-->

```
<div class="markup-2BOw-j">test
<pre><code class="scrollbarGhost-2F9Zj2 scrollbar-3dvm_9 hljs">multiline
code</code></pre>
ing</div>
```

File: test/test_messageprocessor.ts [hunk 4]
```diff
@@ -51,14 +52,14 @@ describe("MessageProcessor", () => {
...
-            Chai.assert.equal(result.formattedBody, "<p>&gt; inb4 tests</p>");
+            Chai.assert.equal(result.formattedBody, "&gt; inb4 tests");
 
             msg.embeds = [];
             msg.content = "[test](http://example.com)";
             result = await processor.FormatDiscordMessage(msg);
             Chai.assert.equal(result.body, "[test](http://example.com)");
             Chai.assert.equal(result.formattedBody,
-                "<p>[test](<a href=\"http://example.com\">http://example.com</a>)</p>");
+                "[test](<a href=\"http://example.com\">http://example.com</a>)");
```
[the remark this answers is not in the document]
Comment (Sorunome) [reply test/test_messageprocessor.ts]: it actualyl is exactly what discord shows.....discord is stupid

File: src/messageprocessor.ts [hunk 5]
```diff
@@ -74,19 +69,21 @@ export class MessageProcessor {
...
         // as else it'll HTML escape the result of the discord syntax
-        let contentPostmark = marked(content).replace(/\n/g, "<br>").replace(/(<br>)?<\/p>(<br>)?/g, "</p>");
+        let contentPostmark = markdown.toHTML(content, {
+            discordCallback: this.getDiscordCallbackHTML(msg),
+        });
 
         // parse the plain text stuff
+        content = markdown.toHTML(content, {
+            discordCallback: this.getDiscordCallback(msg),
+            discordOnly: true,
```
[the remark this answers is not in the document]
Comment (Half-Shot) [reply src/messageprocessor.ts]: Sounds fair actually.

File: test/test_messageprocessor.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (Half-Shot) [reply test/test_messageprocessor.ts]: :cry: fair enough

File: test/test_messageprocessor.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (Half-Shot) [reply test/test_messageprocessor.ts]: Conclusion OOB: discord is awful and this will have to do.

**Pull request merged**